libyul/optimiser/ExpressionInliner.cpp

```diff
@@ -52,21 +51,28 @@ void ExpressionInliner::visit(Expression& _expression)
 	if (_expression.type() == typeid(FunctionCall))
 	{
 		FunctionCall& funCall = boost::get<FunctionCall>(_expression);
+		if (!m_inlinableFunctions.count(funCall.functionName.name))
+			return;
+		FunctionDefinition const& fun = *m_inlinableFunctions.at(funCall.functionName.name);
 
-		bool movable = boost::algorithm::all_of(
-			funCall.arguments,
-			[=](Expression const& _arg) { return SideEffectsCollector(m_dialect, _arg).movable(); }
-		);
-		if (m_inlinableFunctions.count(funCall.functionName.name) && movable)
+		map<YulString, Expression const*> substitutions;
+		for (size_t i = 0; i < funCall.arguments.size(); i++)
 		{
-			FunctionDefinition const& fun = *m_inlinableFunctions.at(funCall.functionName.name);
-			map<YulString, Expression const*> substitutions;
-			for (size_t i = 0; i < fun.parameters.size(); ++i)
-				substitutions[fun.parameters[i].name] = &funCall.arguments[i];
-			_expression = Substitution(substitutions).translate(*boost::get<Assignment>(fun.body.statements.front()).value);
-
-			// TODO Add metric! This metric should perform well on a pair of functions who
-			// call each other recursively.
+			Expression const& arg = funCall.arguments[i];
+			YulString paraName = fun.parameters[i].name;
+
+			if (!SideEffectsCollector(m_dialect, arg).movable())
+				return;
+
+			size_t refs = ReferencesCounter::countReferences(fun.body)[paraName];
+			size_t cost = CodeCost::codeCost(m_dialect, arg);
```

Comment:
can we add an assertion here that `assert(cost != 0 && refs != 0)`

Reply:
Metrics.h says
```
Function calls cost 50. Instructions cost 0 for 3 or less gas (same as DUP),  2 for up to 10 and 50 otherwise.
```
So I think the cost and refs can be zero.

Reply:
Ah, thank you for pointing out. I somehow assumed cost is the same as gas, but it seems it is an abstraction over it that is only used by the optimizer.